### ISSUE TITLE
Create programmable MSBuild Tasks

### ISSUE THREAD

**Issue by hishamco** [OrchardCMS/OrchardCore#4991]
@jtkech I'm thinking to introduce programmable MSBuild tasks that will be useful for various reasons, such as:

- Copy Translations files

- Cleanup files

- or anything you may think of

I knew we can do that in `.csproj` files, be here we can have a benefit of test ability, readability, productivity, flexibility .. etc

**Comment by jtkech** [comment]
yes good idea

**Comment by Piedone** [comment]
Can you tell about some concrete use cases in more detail? Like, what exactly would "Copy Translations files" and "Cleanup files" do?

**Comment by hishamco** [comment]
Working with MSBuild tasks is a tedious work, also a tiny typo could break everything I remembered we using programmable MSBuild tasks in `SimplCommerce` long time ago, this way we can easy create, modify any task

**Comment by Piedone** [comment]
Sorry, this is really vague for me. Please provide examples.

**Comment by hishamco** [comment]
For instance, copying translation files is a simple task that took days if not weeks to make work fine as it is, I remember JT works heavily to get it done.

If we have a programmatic way to create MSBuild task this should not take a day IMHO, for that JT was interested on this at that time

**Comment by Piedone** [comment]
Aren't you looking to [implement custom tasks in C#](https://learn.microsoft.com/en-us/visualstudio/msbuild/task-writing)?

**Comment by hishamco** [comment]
The problem is not with writing the task itself, but rather how it is getting work with OC :)

**Comment by Piedone** [comment]
And how would such programmable tasks work? Wouldn't rather custom MSBuild targets that let you hook into the build pipeline of OC serve this?

**Comment by Piedone** [comment]
So, where do you want to go with this issue and PR?

**Comment by hishamco** [comment]
Hi @Piedone as you know we are on a holy month "Ramadhan" so I don't have much time to revise the remaining PRs. I'll struggle in the upcoming two weeks

**Comment by Piedone** [comment]
I see, no worries, and have a happy Ramadan!

**Comment by hishamco** [comment]
BTW the number of PRs -especially mine - has been reduced a lot, thanks for your `ZoltanGC.Collect()` :)

**Comment by Piedone** [comment]
Great :D.

**Comment by hishamco** [comment]
> And how would such programmable tasks work? Wouldn't rather custom MSBuild targets that let you hook into the build pipeline of OC serve this?

Implementing programmable MSBuild tasks will open up the possibilities for doing anything during the build time:

- Reduce the time for creating a build task from within `.csproj`, I remember JT spend a lot of PR to make copying localization translations works
- Easy to create if you know C# :)
- We can create custom targets or use the built-in once
- PERF nothing involve in the runtime
- We can move almost the stuff that we did in `.csproj` into readable and understandable tasks such as:
1. Copying translation files
2. Creating `wwwroot` folder
3. Creating NuGet packages
4. and much more

The possibilities are endless, you can even create programmable classes for certain end-points or anything you can think of

**Comment by Piedone** [comment]
What I'm probably most missing is what you mean by "programmable" (and even after checking out the PR). Since MSBuild Tasks are already, for all intents and purposes, programmable. Because you previously said that you don't mean being able to write tasks in C#.

**Comment by hishamco** [comment]
> What I'm probably most missing is what you mean by "programmable"

Programmatically

> Since MSBuild Tasks are already, for all intents and purposes, programmable. Because you previously said that you don't mean being able to write tasks in C#.

I can write them in `.csproj` but they are tedious and time-consuming and you find yourself reading the docs all the time :)

I don't mean that we need to change all that we have, at least this PoC that C# tasks are easier and do not take much time. I think this is useful for the future use cases

**Comment by Piedone** [comment]
I understand what the word "programmable" means :). I don't understand though what you mean by it in this context. So, in the end, this is actually about writing MSBuild tasks in C# (and not the MSBuild XML)?

**Comment by hishamco** [comment]
Exactly, open the possibility to write MSBuild tasks in C#, so it's easier, time-saving and testable
